lib/srv/discovery/discovery.go

```diff
@@ -56,46 +83,79 @@ type Server struct {
 	ctx context.Context
 	// cancelfn is used with ctx when stopping the discovery server
 	cancelfn context.CancelFunc
-
-	// log is used for logging.
-	log *logrus.Entry
+	// nodeWatcher is a node watcher.
+	nodeWatcher *services.NodeWatcher
 
 	// cloudWatcher periodically retrieves cloud resources, currently
 	// only EC2
 	cloudWatcher *server.Watcher
 	// ec2Installer is used to start the installation process on discovered EC2 nodes
 	ec2Installer *server.SSMInstaller
+	// kubeFetchers holds all kubernetes fetchers for Azure and other clouds.
+	kubeFetchers []fetchers.Fetcher
 }
 
 // New initializes a discovery Server
 func New(ctx context.Context, cfg *Config) (*Server, error) {
-	if len(cfg.Matchers) == 0 {
-		return nil, trace.NotFound("no matchers present")
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
 	}
 
 	localCtx, cancelfn := context.WithCancel(ctx)
 	s := &Server{
-		Config: cfg,
-		log: logrus.WithFields(logrus.Fields{
-			trace.Component: teleport.ComponentDiscovery,
-		}),
+		Config:   cfg,
 		ctx:      localCtx,
 		cancelfn: cancelfn,
 	}
 
 	var err error
-	s.cloudWatcher, err = server.NewCloudWatcher(s.ctx, s.Matchers, s.Clients)
-	if err != nil {
+	if len(cfg.AWSMatchers) > 0 {
+		s.cloudWatcher, err = server.NewCloudWatcher(s.ctx, cfg.AWSMatchers, s.Clients)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		s.ec2Installer = server.NewSSMInstaller(server.SSMInstallerConfig{
+			Emitter: cfg.Emitter,
+		})
+	}
+
+	for _, matcher := range cfg.AzureMatchers {
+		subscriptions, err := s.getAzureSubscriptions(ctx, matcher.Subscriptions)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, subscription := range subscriptions {
+			for _, t := range matcher.Types {
+				switch t {
+				case constants.AzureServiceTypeKubernetes:
+					kubeClient, err := s.Clients.GetAzureKubernetesClient(subscription)
+					if err != nil {
+						return nil, trace.Wrap(err)
+					}
+					fetcher, err := fetchers.NewAKSFetcher(fetchers.AKSFetcherConfig{
+						Client:         kubeClient,
+						Regions:        matcher.Regions,
+						FilterLabels:   matcher.ResourceTags,
+						ResourceGroups: matcher.ResourceGroups,
+						Log:            s.Log,
+					})
+					if err != nil {
+						return nil, trace.Wrap(err)
+					}
+					s.kubeFetchers = append(s.kubeFetchers, fetcher)
+				}
+			}
+		}
+	}
+
+	if err := s.initTeleportResourceWatchers(s.cloudWatcher != nil); err != nil {
```

Comment:
Instead of passing `s.cloudWatcher != nil` there and exiting right away, can we just not call initTeleportResourceWatchers in that case?